File: app/schemas/analysis.py
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict


class ErrorCode(str, Enum):
    UNSUPPORTED_MEDIA_TYPE = "UNSUPPORTED_MEDIA_TYPE"
    INVALID_IMAGE = "INVALID_IMAGE"
    FILE_TOO_LARGE = "FILE_TOO_LARGE"
    ANALYSIS_FAILED = "ANALYSIS_FAILED"


class ErrorDetail(BaseModel):
    code: ErrorCode
    message: str


class ErrorResponse(BaseModel):
    error: ErrorDetail


class HealthResponse(BaseModel):
    status: str


class DetectionResponse(BaseModel):
    label: str
    score: float
    box: list[float]


class RiskResponse(DetectionResponse):
    model_config = ConfigDict(extra="allow")

    rule_id: str | None = None
    risk_level: str | None = None
    risk_score: int | float | None = None
    reason: str | None = None
    recommendation: str | None = None
    target_group: str | None = None


class AnalysisResponse(BaseModel):
    model_config = ConfigDict(extra="allow")

    image_name: str
    image_path: str | None = None
    detection_count: int
    detections: list[DetectionResponse]
    risk_count: int | None = None
    risks: list[RiskResponse]
    result_image_path: str
    risk_analysis_path: str | None = None
    boxed_image_path: str | None = None


class ApiError(Exception):
    def __init__(
        self,
        status_code: int,
        code: ErrorCode,
        message: str,
    ) -> None:
        self.status_code = status_code
        self.code = code
        self.message = message
        super().__init__(message)


ERROR_RESPONSES: dict[int, dict[str, Any]] = {
    400: {"model": ErrorResponse},
    413: {"model": ErrorResponse},
    500: {"model": ErrorResponse},
}

File: frontend/src/types/analysis.ts
export type SelectedImage = {
  file: File;
  previewUrl: string;
};

export type Detection = {
  label: string;
  raw_label?: string | null;
  canonical_label?: string | null;
  display_label?: string | null;
  score: number;
  box: number[];
};

export type Risk = Detection & {
  rule_id?: string | null;
  risk_level?: string | null;
  risk_score?: number | null;
  reason?: string | null;
  recommendation?: string | null;
  target_group?: string | null;
};

export type AnalysisResponse = {
  image_name: string;
  image_path?: string | null;
  detection_count: number;
  detections: Detection[];
  risk_count?: number | null;
  risks: Risk[];
  result_image_path: string;
  result_image_url?: string | null;
  risk_analysis_path?: string | null;
  boxed_image_path?: string | null;
};

export type AnalysisErrorCode =
  | 'NO_IMAGE_SELECTED'
  | 'API_NOT_CONFIGURED'
  | 'UNSUPPORTED_IMAGE'
  | 'BACKEND_UNREACHABLE'
  | 'BACKEND_4XX'
  | 'BACKEND_5XX'
  | 'MALFORMED_RESPONSE'
  | 'RESULT_IMAGE_UNAVAILABLE';

export type AnalysisError = {
  code: AnalysisErrorCode;
  message: string;
};

export type AnalysisStatus = 'idle' | 'ready' | 'loading' | 'success' | 'error';
